### PR TITLE
[WIP] Robosuite Control

### PR DESCRIPTION
**Pull request by Alacarter**
Adapter and unit tests for Robosuite Environments.

### CODE REVIEW

File: softlearning/environments/adapters/robosuite_adapter.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import gym
+from gym import spaces
+
+from .softlearning_env import SoftlearningEnv
+from softlearning.environments.adapters.gym_adapter import GymAdapter
+
+ROBOSUITE_ENVIRONMENTS = {}
+
+def convert_robosuite_to_gym_obs_space(robosuite_obs_space):
+    assert isinstance(robosuite_obs_space, OrderedDict)
```
Comment (hartikainen) [inline softlearning/environments/adapters/robosuite_adapter.py]:
```python
assert isinstance(robosuite_obs_space, OrderedDict), type(
    robosuite_observation_space)
```

File: softlearning/environments/adapters/robosuite_adapter.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    listDict = []
```
Comment (hartikainen) [inline softlearning/environments/adapters/robosuite_adapter.py]: We usually prefer snake case variable names in python: `list_dict`

File: softlearning/environments/adapters/robosuite_adapter.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    for key, value in robosuite_obs_space.items():
+        listDict.append((key, spaces.Box(
+            low=-float("inf"),
+            high=float("inf"),
+            shape=(len(value),),
```
Comment (hartikainen) [inline softlearning/environments/adapters/robosuite_adapter.py]: What's the value here? Can you use `value.shape`?

File: softlearning/environments/adapters/robosuite_adapter.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+        # TODO(Alacarter): Check how robosuite handles max episode length termination.
+
+        assert isinstance(env.observation_spec(), OrderedDict), env.observation_spec()
+        self.observation_keys = observation_keys or tuple(env.observation_spec().keys())
+        assert set(self.observation_keys).issubset(set(env.observation_spec().keys())), (
+            (self.observation_keys, env.observation_spec().keys()))
+
+        if normalize:
+            np.testing.assert_equal(env.action_spec[0], -1, "Ensure spaces are normalized.")
```
Comment (hartikainen) [inline softlearning/environments/adapters/robosuite_adapter.py]: Could you add a test case for this?


```python
def test_fails_with_unnormalized_action_spec(self):
    env = EnvWithUnnormalizedActionSpec(...)
    with self.assertRaises(AssertionError):
        adapter = RobosuiteAdapter(env, ...)
```

File: tests/softlearning/environments/adapters/robosuite_adapter_test.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                domain=domain,
+                task=task,
+                has_renderer=False,
+                has_offscreen_renderer=False,
+                use_camera_obs=False)
+            env.reset()
+            env.step(env.action_space.sample())
+
+        for domain, task in TEST_ENVIRONMENTS:
+            print("testing: ", domain, task)
```
Comment (hartikainen) [inline tests/softlearning/environments/adapters/robosuite_adapter_test.py]: Remove

File: tests/softlearning/environments/adapters/robosuite_adapter_test.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        env2 = env1.copy()
+
+        self.assertEqual(env1.observation_keys, env2.observation_keys)
+        for key, value in env_kwargs.items():
+            self.assertEqual(getattr(env1.unwrapped, key), value)
+            self.assertEqual(getattr(env2.unwrapped, key), value)
+
+        domain, task = 'Sawyer', 'Lift'
+        robosuite_adapter_kwargs = {
+            'observation_keys': ('joint_pos', 'joint_vel')
```
Comment (hartikainen) [inline tests/softlearning/environments/adapters/robosuite_adapter_test.py]: This used to pass with incorrect `observation_keys`. Can you add a test that makes sure that incorrect observation keys fail?

File: softlearning/environments/adapters/robosuite_adapter.py (hunk 2 above)
Comment (hartikainen) [reply softlearning/environments/adapters/robosuite_adapter.py]: Btw, Python Enhancement Proposal (PEP 8 specifically) provides official guidelines and for Python styles in general: https://www.python.org/dev/peps/pep-0008/.

I'm not saying that you should read that, but it's good to keep in mind. Linter will do all this automatically for you though :smile:

File: softlearning/environments/adapters/robosuite_adapter.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+def convert_robosuite_to_gym_obs_space(robosuite_obs_space):
+    assert isinstance(robosuite_obs_space, OrderedDict)
+    listDict = []
+    for key, value in robosuite_obs_space.items():
+        listDict.append((key, spaces.Box(
+            low=-float("inf"),
+            high=float("inf"),
+            shape=(len(value),),
+            dtype=np.float32)))
```
Comment (hartikainen) [inline softlearning/environments/adapters/robosuite_adapter.py]: Maybe use `value.dtype` here if possible?

File: softlearning/environments/adapters/robosuite_adapter.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    return spaces.Dict(OrderedDict(listDict))
+
+def convert_robosuite_to_gym_action_space(robosuite_action_space):
+    assert isinstance(robosuite_action_space, tuple)
```
Comment (hartikainen) [inline softlearning/environments/adapters/robosuite_adapter.py]:
```suggestion
    assert isinstance(robosuite_action_space, tuple), type(robosuite_action_space)
```

File: softlearning/environments/adapters/robosuite_adapter.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    """Adapter that implements the SoftlearningEnv for Robosuite envs."""
+
+    def __init__(self,
+                 domain,
+                 task,
+                 *args,
+                 env=None,
+                 normalize=True,
+                 observation_keys=None,
+                 unwrap_time_limit=True,
```
Comment (hartikainen) [inline softlearning/environments/adapters/robosuite_adapter.py]: I don't think robosuite wraps their environments, so we can remove this argument.

**Comment by hartikainen** [comment]
Generally looking good! I left a few minor comments in the code. Can you ping me here when this is ready for another review?

File: softlearning/environments/adapters/robosuite_adapter.py (hunk 3 above)
Comment (Alacarter) [reply softlearning/environments/adapters/robosuite_adapter.py]: Yeah that's a much better idea!

File: softlearning/environments/adapters/robosuite_adapter.py (hunk 4 above)
Comment (Alacarter) [reply softlearning/environments/adapters/robosuite_adapter.py]: Hmm I'm not too sure how to get an environment that isn't normalized. I tried modifying the `action_spec` attribute of an existing instance of an environment to denormalize it, but I'm not able to set attribute values.

`>>> envs = ["SawyerLift", "SawyerStack", "SawyerPickPlace", "SawyerPickPlaceSingle", "SawyerPickPlaceMilk", "SawyerPickPlaceBread", "SawyerPickPlaceCereal", "SawyerPickPlaceCan", "SawyerNutAssembly", "SawyerNutAssemblySingle", "SawyerNutAssemblySquare", "SawyerNutAssemblyRound", "BaxterLift", "BaxterPegInHole"]
>>> for env_id in envs:
...     env = suite.make(env_id, has_renderer=False, has_offscreen_renderer=False, use_camera_obs=False)
...     print(env_id, env.action_spec)
... 
SawyerLift (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerStack (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlace (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceSingle (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceMilk (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceBread (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceCereal (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceCan (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerNutAssembly (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerNutAssemblySingle (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerNutAssemblySquare (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerNutAssemblyRound (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
BaxterLift (array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]))
BaxterPegInHole (array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1.]), array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]))
`

File: tests/softlearning/environments/adapters/robosuite_adapter_test.py (hunk 6 above)
Comment (Alacarter) [reply tests/softlearning/environments/adapters/robosuite_adapter_test.py]: Yep, implemented!

File: softlearning/environments/adapters/robosuite_adapter.py (hunk 7 above)
Comment (Alacarter) [reply softlearning/environments/adapters/robosuite_adapter.py]: Yep it works!

File: softlearning/environments/adapters/robosuite_adapter.py (hunk 4 above)
Comment (hartikainen) [reply softlearning/environments/adapters/robosuite_adapter.py]: The way you would test this is by creating your own *minimal* RobosuiteEnv. Something like:

```
from robosuite.environments import MujocoEnv

class MockEnv(MujocoEnv):
    # Fill in the methods that are required for minimal environment.
    @property
    def dof(self):
        return 5

    @property
    def action_spec(self):
        low, high = np.ones(self.dof) * -2.0, np.ones(self.dof) * 2.0
        return low, high
```

I'm not 100% sure what else is required for creating a minimal environment like this. Just try creating a test for an environment like this and see where it fails. Let me know if there's something you get stuck in.

File: tests/softlearning/environments/adapters/softlearning_env_test.py [hunk 10]
```diff
@@ -18,7 +18,7 @@ def test_active_observation_shape(self):
         env = self.create_adapter()
         active_observation_shape = env.active_observation_shape
         self.assertEqual(len(active_observation_shape), 1)
-        self.assertIsInstance(active_observation_shape[0], int)
+        self.assertTrue(np.issubdtype(type(active_observation_shape[0]), np.int64))
```
Comment (hartikainen) [inline tests/softlearning/environments/adapters/softlearning_env_test.py]: Should the second argument be `int` or `np.int`?

Comment (Alacarter) [reply tests/softlearning/environments/adapters/softlearning_env_test.py]: It seems like `int`, `np.int`, and `np.int64` all work for both the dm_control tests and the softlearning tests. So I left it as `np.int64`.

Comment (hartikainen) [reply tests/softlearning/environments/adapters/softlearning_env_test.py]: Let's make it `np.int`, I think that's a bit more general.

**Comment by hartikainen** [comment]
Awesome, this is starting to look good. Have you tested that the environments run with SAC?

File: examples/development/variants.py [hunk 11]
```diff
@@ -189,6 +189,13 @@
         'Wall-v0': {
             'observation_keys': ('observation', 'desired_goal'),
         },
+    },
+    'Sawyer': {
+      'Lift': {
```
Comment (hartikainen) [inline examples/development/variants.py]: Indentation

Comment (Alacarter) [reply examples/development/variants.py]: My bad, should be fixed now.

**Comment by hartikainen** [comment]
I made some minor syntax fixes and cleanups and merged master to this branch. I tested all the Sawyer environments and they run fine. None of the environments were solved perfectly by SAC, but I think that's more of an algorithm issue and I think the environments work as expected.

I'll merge this. Good job @Alacarter, thanks for adding this!

File: softlearning/environments/utils.py [hunk 12]
```diff
@@ -8,15 +8,22 @@
     DmControlAdapter,
 )
 
+from .adapters.robosuite_adapter import (
```
Comment (hartikainen) [inline softlearning/environments/utils.py]: One thing to do here is that we need to make all these adapters fail lazily if they can't be imported, such that not all the different environments need to be installed.

**Comment by hartikainen** [comment]
Actually, before merging this, I think we need to make sure that `robosuite` and `dm_control` installations are optional, and we fail lazily if they are used but not installed. See my comment in line.

**Comment by hartikainen** [comment]
Ok, I made the imports fail lazily. I'll merge this.

**Pull request merged**